Neos.Neos/Classes/AssetUsage/CatchUpHook/AssetUsageCatchUpHook.php

```diff
@@ -0,0 +1,156 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Neos\Neos\AssetUsage\CatchUpHook;
+
+use Neos\ContentRepository\Core\ContentRepository;
+use Neos\ContentRepository\Core\DimensionSpace\DimensionSpacePoint;
+use Neos\ContentRepository\Core\DimensionSpace\DimensionSpacePointSet;
+use Neos\ContentRepository\Core\EventStore\EventInterface;
+use Neos\ContentRepository\Core\Feature\Common\EmbedsContentStreamId;
+use Neos\ContentRepository\Core\Feature\Common\EmbedsWorkspaceName;
+use Neos\ContentRepository\Core\Feature\DimensionSpaceAdjustment\Event\DimensionSpacePointWasMoved;
+use Neos\ContentRepository\Core\Feature\NodeCreation\Event\NodeAggregateWithNodeWasCreated;
+use Neos\ContentRepository\Core\Feature\NodeModification\Event\NodePropertiesWereSet;
+use Neos\ContentRepository\Core\Feature\NodeRemoval\Event\NodeAggregateWasRemoved;
+use Neos\ContentRepository\Core\Feature\NodeVariation\Event\NodeGeneralizationVariantWasCreated;
+use Neos\ContentRepository\Core\Feature\NodeVariation\Event\NodePeerVariantWasCreated;
+use Neos\ContentRepository\Core\Feature\NodeVariation\Event\NodeSpecializationVariantWasCreated;
+use Neos\ContentRepository\Core\Feature\WorkspacePublication\Dto\NodeIdsToPublishOrDiscard;
+use Neos\ContentRepository\Core\Feature\WorkspacePublication\Event\WorkspaceWasDiscarded;
+use Neos\ContentRepository\Core\Feature\WorkspacePublication\Event\WorkspaceWasPartiallyDiscarded;
+use Neos\ContentRepository\Core\Projection\CatchUpHookInterface;
+use Neos\ContentRepository\Core\Projection\ContentGraph\Filter\FindDescendantNodesFilter;
+use Neos\ContentRepository\Core\Projection\ContentGraph\Node;
+use Neos\ContentRepository\Core\Projection\ContentGraph\VisibilityConstraints;
+use Neos\ContentRepository\Core\SharedModel\Node\NodeAggregateId;
+use Neos\ContentRepository\Core\SharedModel\Workspace\WorkspaceName;
+use Neos\EventStore\Model\EventEnvelope;
+use Neos\Neos\AssetUsage\Service\AssetUsageIndexingService;
+
+/**
+ * @internal
+ */
+class AssetUsageCatchUpHook implements CatchUpHookInterface
+{
+    public function __construct(
+        private readonly ContentRepository $contentRepository,
+        private readonly AssetUsageIndexingService $assetUsageIndexingService
+    ) {
+    }
+
+    public function onBeforeCatchUp(): void
+    {
+    }
+
+    public function onBeforeEvent(EventInterface $eventInstance, EventEnvelope $eventEnvelope): void
+    {
+        if (
+            $eventInstance instanceof EmbedsWorkspaceName
+            && $eventInstance instanceof EmbedsContentStreamId
+            // Safeguard for temporary content streams created during partial publish -> We want to skip these events, because their workspace doesn't match current contentstream.
+            && !$this->contentRepository->getWorkspaceFinder()->findOneByCurrentContentStreamId($eventInstance->getContentStreamId())?->workspaceName->equals($eventInstance->getWorkspaceName())
+        ) {
+            return;
+        }
```

Comment:
@dlubitz do you agree?
```suggestion
        if ($eventInstance instanceof EmbedsWorkspaceName && $eventInstance instanceof EmbedsContentStreamId) {
            // Safeguard for temporary content streams created during partial publish -> We want to skip these events, because their workspace doesn't match current content stream.
            try {
                $contentGraph = $this->contentRepository->getContentGraph($eventInstance->getWorkspaceName());
            } catch (WorkspaceDoesNotExist) {
                return:
            }
            if (!$contentGraph->getContentStreamId()->equals($eventInstance->getContentStreamId()) {
                return;
            }            
        }
```

BTW: Wasn't the plan to use some dedicated virtual workspace name for those temporary streams?

Reply:
Yes this was the plan, but was not scope of this issue. This change just raised that.

Regarding the suggested change: Yeah, I'm fine with both as this will be removed/replaced anyway. Will change it later.

Reply:
Done